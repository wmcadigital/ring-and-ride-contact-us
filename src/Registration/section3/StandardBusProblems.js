import { useFormState } from "react-final-form";

import FormSection from "../../common/FormSection";
import ProgressIndicator from "../../common/ProgressIndicator";
import Question from "../../common/Question";
import getRequirementsSectionName from "./getRequirementsSectionName";
import CheckboxContainer from "../../common/CheckboxContainer";
import Checkbox from "../../common/Checkbox";
import FieldError from "../../common/FieldError";

export const BusProblemOptions = (registerForYourself) => {
  if (registerForYourself) {
    return {
      cannotGetToBus: "I cannot get to a bus stop",
      cannotGetOnBus: "I cannot get on a bus",
      cannotTravelOnBus: "I cannot travel on a bus",
    };
  }
  return {
    cannotGetToBus: "They cannot get to a bus stop",
    cannotGetOnBus: "They cannot get on a bus",
    cannotTravelOnBus: "They cannot travel on a bus",
  };
};

const StandardBusProblems = () => {
  const stateApi = useFormState();

  const formValues = stateApi.values;
  const registerForYourself = formValues["registerForYourself"] === "yes";
  const question = registerForYourself
    ? "What problems do you experience when trying to use standard bus services?"
    : `What problems does ${formValues["firstName"]} experience when trying to use standard bus services?`;

  const standardBusProbError = stateApi.submitFailed
    ? stateApi.errors?.selectStandardBusProb
    : null;

  const busProblemOptions = BusProblemOptions(registerForYourself);

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition="Section 3 of 3"
        sectionName={getRequirementsSectionName(stateApi)}
      />
      <Question text={question} />
      <p>Select all that apply</p>
      <CheckboxContainer error={standardBusProbError}>
        <FieldError text={standardBusProbError} />
        {Object.keys(busProblemOptions).map((option) => (
          <Checkbox
            key={busProblemOptions[option]}
            label={busProblemOptions[option]}
            fieldName={`standardBusProb.${option}`}
          />
        ))}
      </CheckboxContainer>
    </FormSection>
  );
};

export default StandardBusProblems;
