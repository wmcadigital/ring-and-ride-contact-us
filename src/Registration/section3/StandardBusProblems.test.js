import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import StandardBusProblems from "./StandardBusProblems";

jest.mock("react-final-form");

describe("StandardBusProblem", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusProblems />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "no",
        firstName: "Test",
        lastName: "Testerson",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusProblems />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
      submitFailed: true,
      errors: { selectStandardBusProb: "Select at least 1 topic" },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusProblems />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
