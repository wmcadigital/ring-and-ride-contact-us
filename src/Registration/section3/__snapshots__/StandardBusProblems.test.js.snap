// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardBusProblem individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Applicant requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="What problems does Test experience when trying to use standard bus services?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="standardBusProb.cannotGetToBus"
      label="They cannot get to a bus stop"
    />
    <Checkbox
      fieldName="standardBusProb.cannotGetOnBus"
      label="They cannot get on a bus"
    />
    <Checkbox
      fieldName="standardBusProb.cannotTravelOnBus"
      label="They cannot travel on a bus"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`StandardBusProblem individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="What problems do you experience when trying to use standard bus services?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="standardBusProb.cannotGetToBus"
      label="I cannot get to a bus stop"
    />
    <Checkbox
      fieldName="standardBusProb.cannotGetOnBus"
      label="I cannot get on a bus"
    />
    <Checkbox
      fieldName="standardBusProb.cannotTravelOnBus"
      label="I cannot travel on a bus"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`StandardBusProblem renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="What problems do you experience when trying to use standard bus services?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error="Select at least 1 topic"
  >
    <FieldError
      text="Select at least 1 topic"
    />
    <Checkbox
      fieldName="standardBusProb.cannotGetToBus"
      label="I cannot get to a bus stop"
    />
    <Checkbox
      fieldName="standardBusProb.cannotGetOnBus"
      label="I cannot get on a bus"
    />
    <Checkbox
      fieldName="standardBusProb.cannotTravelOnBus"
      label="I cannot travel on a bus"
    />
  </CheckboxContainer>
</FormSection>
`;
